contracts/src/token/erc721/extensions/consecutive.rs

```diff
@@ -0,0 +1,1079 @@
+/// stress on off-chain indexing services that have to record one entry per
+/// token, and have protections against "unreasonably large" batches of tokens.
+pub const MAX_BATCH_SIZE: U96 = uint!(5000_U96);
+
+/// Used to offset the first token id in
+/// [`Erc721Consecutive::_next_consecutive_id`]
+pub const FIRST_CONSECUTIVE_ID: U96 = uint!(0_U96);
+
+/// Consecutive extension related implementation:
+impl Erc721Consecutive {
+    /// Override of [`Erc721::_owner_of_inner`] that checks the sequential
+    /// ownership structure for tokens that have been minted as part of a
+    /// batch, and not yet transferred.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    /// * `token_id` - Token id as a number.
+    pub fn _owner_of_inner(&self, token_id: U256) -> Address {
+        let owner = self.__owner_of_inner(token_id);
+        // If token is owned by the core, or beyond consecutive range, return
+        // base value
+        if owner != Address::ZERO
+            || token_id > U256::from(U96::MAX)
+            || token_id > U256::from(U96::MAX)
+            || token_id < U256::from(FIRST_CONSECUTIVE_ID)
+        {
+            return owner;
+        }
+
+        // Otherwise, check the token was not burned, and fetch ownership from
+        // the anchors.
+        // NOTE: no need for safe cast,
+        if self._sequentian_burn.get(token_id) {
+            Address::ZERO
+        } else {
+            self._sequential_ownership.lower_lookup(U96::from(token_id)).into()
+        }
+    }
+
+    /// Mint a batch of tokens of length `batch_size` for `to`. Returns the
+    /// token id of the first token minted in the batch; if `batchSize` is
+    /// 0, returns the number of consecutive ids minted so far.
+    ///
+    /// Requirements:
+    ///
+    /// - `batchSize` must not be greater than [`MAX_BATCH_SIZE`].
+    /// - The function is called in the constructor of the contract (directly or
+    ///   indirectly).
+    ///
+    /// CAUTION: Does not emit a `Transfer` event. This is ERC-721 compliant as
+    /// long as it is done inside of the constructor, which is enforced by
+    /// this function.
+    ///
+    /// CAUTION: Does not invoke `onERC721Received` on the receiver.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Write access to the contract's state.
+    /// * `token_id` - Token id as a number.
+    ///
+    /// # Errors
+    ///
+    /// If to is [`Address::ZERO`] error [`rc721::Error::InvalidReceiver`] is
+    /// returned.
+    /// If batch size exceeds [`MAX_BATCH_SIZE`] error
+    /// [`Error::ERC721ExceededMaxBatchMint`] is returned.
+    ///
+    /// # Events
+    ///
+    /// Emits a [`ConsecutiveTransfer`] event.
+    pub fn _mint_consecutive(
+        &mut self,
+        to: Address,
+        batch_size: U96,
+    ) -> Result<U96, Error> {
+        let next = self._next_consecutive_id();
+
+        if batch_size > U96::ZERO {
```

Comment:
Let's keep the comments from the original implementation with proper capitalization and ending dots:

```solidity
        // minting a batch of size 0 is a no-op
        if (batchSize > 0) {
            if (address(this).code.length > 0) {
                revert ERC721ForbiddenBatchMint();
            }
            if (to == address(0)) {
                revert ERC721InvalidReceiver(address(0));
            }

            uint256 maxBatchSize = _maxBatchSize();
            if (batchSize > maxBatchSize) {
                revert ERC721ExceededMaxBatchMint(batchSize, maxBatchSize);
            }

            // push an ownership checkpoint & emit event
            uint96 last = next + batchSize - 1;
            _sequentialOwnership.push(last, uint160(to));

            // The invariant required by this function is preserved because the new sequentialOwnership checkpoint
            // is attributing ownership of `batchSize` new tokens to account `to`.
            _increaseBalance(to, batchSize);

            emit ConsecutiveTransfer(next, last, address(0), to);
        }
```